timeutil/timer.go

```diff
@@ -0,0 +1,178 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package timer provides a wrapper around time.Timer and its associated
+// functions so that timers can be mocked in tests.
+package timeutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+// Timer is an interface which wraps time.Timer so that it may be mocked.
+type Timer interface {
+	Reset(d time.Duration) bool
+	Stop() bool
+}
+
+// FakeAfterFunc creates a new fake timer which will call the given callback in
+// its own goroutine when the timer fires. The returned FakeTimer's C field is
+// not used and will be nil.
+//
+// This simulates the behavior of AfterFunc() from the time package.
+// See here for more details: https://pkg.go.dev/time#AfterFunc
+func FakeAfterFunc(d time.Duration, f func()) *FakeTimer {
```

Comment:
this is a test helper only so it could probably live under testutils or timeutil/timeutiltest ?